libraries/standard/mqtt/src/mqtt.c

```diff
@@ -130,6 +234,447 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+    uint16_t packetIdentifier;
+    /* Need a dummy variable for MQTT_DeserializeAck(). */
+    bool sessionPresent = false;
+    MQTTPubAckType_t ackType;
+
+    assert( pContext != NULL );
+    assert( pIncomingPacket != NULL );
+
+    switch( pIncomingPacket->type )
+    {
+        case MQTT_PACKET_TYPE_PUBACK:
+        case MQTT_PACKET_TYPE_PUBREC:
+        case MQTT_PACKET_TYPE_PUBREL:
+        case MQTT_PACKET_TYPE_PUBCOMP:
+            ackType = getAckFromPacketType( pIncomingPacket->type );
+            status = MQTT_DeserializeAck( pIncomingPacket, &packetIdentifier,  &sessionPresent );
+            LogInfoWithArgs( "Ack packet deserialized with result: %d.", status );
+            if( status == MQTTSuccess )
+            {
+                publishRecordState = MQTT_UpdateStateAck( pContext,
+                                                          packetIdentifier,
+                                                          ackType,
+                                                          MQTT_RECEIVE );
+                LogInfoWithArgs( "State record updated. New state=%d.",
+                                 publishRecordState );
+
+                /* Send PUBREL or PUBCOMP if necessary. */
+                status = sendPublishAcks( pContext,
+                                          packetIdentifier,
+                                          &publishRecordState );
+                if( status == MQTTSuccess )
+                {
+                    pContext->callbacks.appCallback( pContext,
+                                                     pIncomingPacket,
+                                                     packetIdentifier,
+                                                     NULL );
+                }
+            }
+            break;
+        case MQTT_PACKET_TYPE_PINGRESP:
+            pContext->waitingForPingResp = false;
+            status = MQTT_DeserializeAck( pIncomingPacket, &packetIdentifier, &sessionPresent );
+            if( status == MQTTSuccess )
+            {
+                pContext->callbacks.appCallback( pContext,
+                                                 pIncomingPacket,
+                                                 packetIdentifier,
+                                                 NULL );
+            }
+            break;
+        case MQTT_PACKET_TYPE_SUBACK:
+        case MQTT_PACKET_TYPE_UNSUBACK:
+            /* Deserialize and give these to the app provided callback. */
+            status = MQTT_DeserializeAck( pIncomingPacket, &packetIdentifier, &sessionPresent );
+            if( status == MQTTSuccess )
+            {
+                pContext->callbacks.appCallback( pContext,
+                                                 pIncomingPacket,
+                                                 packetIdentifier,
+                                                 NULL );
+            }
+            break;
+        default:
+            /* Not an ack. */
```

Comment:
Same comment as in `sendPublishAcks()`, since we're dealing with input from the network here be very clear what is possible/expected here.  Why would I ever be in a function called `handleIncomingAck()` for something which is `"Not an ack"`?   In network packet handling code, we'd usually expect one of three things in a `default:`

1. `assert()` for programming errors
1. log/return failure values for run-time ones
1. _detailed_ comments explaining why neither option above applies.

Reply:
Good catch. The only packet type the server should send is a PUBLISH (handled in another function, or one of the acks above. Anything else is a bad response. I will add an error log.